Added subdomain names and point locations to TopResidualDebugOutput

joshuahansel opened the pull request:
Closes #11676

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

joshuahansel committed:
Added subdomain names and point locations to TopResidualDebugOutput

Closes #11676

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/200700/) on 83b8ad0 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11679/site)

This comment will be updated on new commits.

permcody commented on framework/include/outputs/TopResidualDebugOutput.h:
```diff
@@ -28,17 +28,29 @@ InputParameters validParams<TopResidualDebugOutput>();
+  }
 
   TopResidualDebugOutputTopResidualData(unsigned int var,
+                                        std::set<SubdomainID> subdomain_ids,
```
Pass this by const ref.

permcody commented on framework/include/outputs/TopResidualDebugOutput.h:
```diff
@@ -28,17 +28,29 @@ InputParameters validParams<TopResidualDebugOutput>();
   TopResidualDebugOutputTopResidualData(unsigned int var,
+                                        std::set<SubdomainID> subdomain_ids,
                                         dof_id_type nd,
+                                        Point point,
```
Pass this by const ref

permcody commented on framework/src/outputs/TopResidualDebugOutput.C:
```diff
@@ -55,23 +55,24 @@ TopResidualDebugOutput::output(const ExecFlagType & /*type*/)
         dof_id_type dof_idx = node->dof_number(_sys.number(), var, 0);
-        vec[j] = TopResidualDebugOutputTopResidualData(var, nd, residual(dof_idx));
+        vec[j] = TopResidualDebugOutputTopResidualData(
+            var, subdomain_ids, nd, Point(*node), residual(dof_idx));
```
Once you switch to const ref in the declaration, you won't have to create a temporary point here. You can just pass in the *node directly.

permcody commented on framework/src/outputs/TopResidualDebugOutput.C:
```diff
@@ -89,7 +90,8 @@ TopResidualDebugOutput::printTopResiduals(const NumericVector<Number> & residual
         {
-          vec[j] = TopResidualDebugOutputTopResidualData(var_num, 0, residual(dof), true);
+          vec[j] =
+              TopResidualDebugOutputTopResidualData(var_num, {}, 0, Point(), residual(dof), true);
```
Hmm... Now this call will cause a problem with the const ref to a set. I guess you can leave it as a copy unless you want to add another override.

joshuahansel replied:
I didn't think I could store any constant references in the structure because of this no-argument constructor getting called in `vector<T>::resize`. As far as *passing* by constant reference, would I actually gain anything since the data immediately gets copied? I thought that the compiler would optimize out the double copy.

permcody replied:
Yeah, if this is your use case, then don't worry about the const ref. The compiler likely will optimize out the double copy and since this isn't in a loop. You never know thought, but we don't need to prematurely over-optimize it.

permcody commented on framework/src/outputs/TopResidualDebugOutput.C:
```diff
@@ -113,6 +115,23 @@ TopResidualDebugOutput::printTopResiduals(const NumericVector<Number> & residual
+          prepend_comma = true;
+        subdomains_ss << "'" << mesh.getSubdomainName(subdomain_id) << "'";
+      }
+      subdomains_ss << "}";
```
This whole block of code is unneeded, use Moose::stringify() and pass in your container. It should work on this an pretty-print it for you.

joshuahansel replied:
`Moose::stringify()` sort-of pretty-prints it: I get "block_one,block_two" instead of what I had before, "{'block_one', 'block_two'}". I'd argue that mine is prettier. At a minimum, for clarity in the output, I'd like containing braces for the string, so I'd still need to append those. Optionally, I think the quotes make it more clear that they are separate items, but I could live without them. Ideally, there'd be a `Moose::prettyPrintVector()` or something like this (or altering the `stringify` method, but I know that would break things). How should I proceed here?

joshuahansel replied:
I agree, I'll see if anything breaks by making `stringify` "prettier".

joshuahansel committed:
Generalized container stringify to print prettier

Added option to enclose the list in curly braces, and added
user-specified string to enclose each element, such as a quotation
mark.
